cudapoa/src/cudapoa_batch.cpp

```diff
@@ -312,6 +312,83 @@ StatusType CudapoaBatch::get_msa(std::vector<std::vector<std::string>>& msa, std
     return StatusType::success;
 }
 
+void CudapoaBatch::get_graphs(std::vector<DirectedGraph>& graphs, std::vector<StatusType>& output_status)
+{
+    int32_t max_nodes_per_window_ = banded_alignment_ ? CUDAPOA_MAX_NODES_PER_WINDOW_BANDED : CUDAPOA_MAX_NODES_PER_WINDOW;
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->nodes,
+                                     graph_details_d_->nodes,
+                                     sizeof(uint8_t) * max_nodes_per_window_ * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edges,
+                                     graph_details_d_->incoming_edges,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * CUDAPOA_MAX_NODE_EDGES * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edge_weights,
+                                     graph_details_d_->incoming_edge_weights,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * CUDAPOA_MAX_NODE_EDGES * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edge_count,
+                                     graph_details_d_->incoming_edge_count,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(input_details_h_->sequence_lengths,
+                                     input_details_d_->sequence_lengths,
+                                     global_sequence_idx_ * sizeof(uint16_t),
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(output_details_h_->consensus,
+                                     output_details_d_->consensus,
+                                     CUDAPOA_MAX_CONSENSUS_SIZE * max_poas_ * sizeof(uint8_t),
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaStreamSynchronize(stream_));
+
```

Comment:
Since you know the final number of graphs, and copying the graphs will be rather expensive, you could do a `graphs.reserve(poa_count_);` here.

Reply:
Right now it's an `emplace_back`, but I can overlap the reserve call with the cuda memcpies to use that latency

Reply:
`emplace_back()` will only place the newly created object directly in the vector (which I think it doesn't even do if you create the object explicitly. I think you should do `v.emplace_back({});`, such that the constructor call gets forwarded).
The problem I see is that the vector will grow in each iteration, which implies the vector will reallocate its memory every now and then (when `v.size()` reached `v.capacity()`). This implies it has to move (or copy if it can't move) all its current elements to the newly allocated memory. So I would like to set the vector's capacity to the expected size right from the start by calling `v.reserve(expected_size)`. The `emplace_back()` and everything else will remain the same.

Reply:
Replaced with `resize` which will allocate and initialize together

Reply:
That's probably even better.